doc/otp-ov-plugin.1
...\" $Header: /usr/src/docbook-to-man/cmd/RCS/docbook-to-man.sh,v 1.3 1996/06/17 03:36:49 fld Exp $
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
\\&.
.fl			\" force out current output buffer
\\!%PB
\\!/showpage{}def
...\" the following is from Ken Flowers -- it prevents dictionary overflows
\\!/tempdict 200 dict def tempdict begin
.fl			\" prolog
.sy cat \\$1\" bring in postscript file
...\" the following line matches the tempdict above
\\!end % tempdict %
\\!PE
\\!.
.sp \\$2u	\" move below the image
..
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n 
.TH "\fBotp-ov-plugin\fP" "1"
.SH "NAME"
\fBotp-ov-plugin\fP \(em OpenVPN plug-in authentication module for OTP database\&.
.SH "SYNOPSIS"
.PP
\fBotp-ov-plugin\fP [-?hv]  [-o\fI otpdb_fname\fP]  [-V\fI service_name\fP]  [-w\fI otp_window\fP] 
.SH "DESCRIPTION"
.PP
The \fBotp-ov-plugin\fP command is plug-in authentication
module for OpenVPN authenticating to the OTP database\&.  The username
and password (challenge response) are passed via environment variables
\fIuser\fP and \fIpass\fP\&.
A successful authentication will set a return code of
\fB0\fR, authentication failure
\fB1\fR, and program failure
\fB-1\fR\&.
.SH "OPTIONS"
.IP "-h, --help" 10
Help
.IP "-o, --otp-db=\fI otpdb_pathname\fP" 10
Pathname of OTP database\&.
.IP "-u, --otp-allow-unknown-user" 10
Allow users which do not exist in the OTP database to successfully
authenticate without using a One Time Password\&.
.IP "-V, --service-name=\fI service_name\fP" 10
Set service name for send-token function\&.
.IP "-v, --verbose" 10
Verbose
.IP "-w, --otp-challenge-window=" 10
Set the OTP challenge window\&.
.IP "--version" 10
Display software version\&.
.SH "EXAMPLES"
.PP
Test the module with user bob\&.
.PP
  \fBexport user="bob"\fP
.PP
  \fBexport pass="882B0E8410"\fP
.PP
  \fBotp-ov-plugin; echo $?\fP
.PP
.nf
1
.fi
.SH "AUTHOR"
.PP
Mark Fullmer maf@splintered\&.net
.SH "SEE ALSO"
.PP
\fBotp-sca\fP(1)
\fBotp-sct\fP(1)
\fBotp-control\fP(1)
\fBpam_otp\fP(1)
\fBhtsoft-downloader\fP(1)
\fBurd\fP(1)
\fBbcload\fP(1)
\fBOpenVPN\fP(8)
spyrus-par2(7)
...\" created by instant / docbook-to-man, Thu 12 Dec 2013, 10:40


doc/CHANGES
# $Id:$

htsoft-downloader: Makefile install target fix - eshicks

pam_otp: allown_unknown typo - eshicks

otplib: more sanity checking on otp_user_add - eshicks

otp-tokend: smtp support, fork() on message

urd/rc.d/urd: updated for new flags

otp-control: mode generate will honor -c

send-tokendd: reference implementation of OOB token.  Sends token to web
page via ssl.  Used internally for SMS gateway.

send-token feature.  Send to token to user via flexible OOB method.

HOTPC.BAS rev 9.  Check PIN, readerKey

otp-sca support SC_SETREADERKEY_CAP

pam_otp.so: require_db_entry is now defaulted to enabled.  A new option
allow_unknown_user which is mutually exclusive to require_db_entry is
accepted for consistency with urd and openvpn.  kyriacou@osc.edu

otpdb2sc missing quote.  kyriacou@osc.edu

openvpn plugin accepts otp-allow-unknown-user

otp-control will by default create a new user with count=1.  BasicCard
code treats count=0 as a special case when a challenge (count) is supplied.
0 is ignore challenge and use EEPROM stored count.

getopt_long() support.  All binaries support --version.

honor OTP_FLAGS_DISPNT in urd.  Use -c to force display

urd will omit password when packed decode debug is enabled

fix state hex encoding in urd when state > 127

fix soft reload option in urd

add PAM password authentication to urd

str.c: tcsetattr() missing in str_input()

otp-control -m list-sc will automatically set the correct FMT flags

document behavior of count=0 in otp-sct, otp-sca, and spyrus-reader

add urd/rc.d script to dist

add rfc.card to dist

spyrus1.4.hex installed instead of 1.3.  Also in QUICKSTART

### ootp-1.03.tar.gz

add CHANGES and TODO to dist

balance card enabled by default with HOTPC

finish dec31.x and DHEX40 otp-sca, otp-sct

add -i option to htsoft-downloader to ignore last WOK timeout

reorganize BasicCard code to minimize RAM and Flash resources.  Fix
PRDisplay (balance reader) syntax error intruduced with BalanceCardIndex

### snap-12-16-2009

add dhex40 format

add set-flags

remove activate, disable, deactivate, flags-dspcnt, flags-no-dspcnt
from otp-control.  Add more generic set-flags, set-format, set-status,
and set-type.

add dec31.6 dec31.7 dec31.8 dec31.9 dec31.10 formats.

add runtime window option to otp-openvpn, urd, and pam_otp.

scr: default to embedded reader if pcscd is not running.  PCSC: defaults
 to first PCSC reader.

urd: -O not positionally dependent on -o. Can disable OTP and specify otpdb.

add FreeBSD rc.d script for urd

add Zeitcontrol balance reader datasheet to docs

### ootp-1.02.tar.gz

otpdb2sc: do not right 0 pad index to otp-control

pam_otp.so: compile PIC

### ootp-1.01.tar.gz

release: build man and html pages at release time

sccmd.c: SetAdminKey fw version check reversed

urd: build default HOTP enabled

### ootp-1.00.tar.gz


otp-openvpn/otp-ov-plugin.c
/*
 * Copyright (c) 2005 Mark Fullmer
 * Copyright (c) 2009 Mark Fullmer and the Ohio State University
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *      $Id: otp-ov-plugin.c 177 2011-05-16 02:37:28Z maf $
 */

#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/errno.h>
#include "ffdb.h"
#include "otplib.h"
#include "xerr.h"

void help(void);

int main (int argc, char **argv)
{
  extern char *ootp_version;
  struct otp_ctx *otpctx;
  struct otp_user ou;
  u_long tmpul;
  char *otpdb_fname, *username, *pass, *endptr, *service;
  int db_flags, i, r, ret, otp_window, opt_version, otp_allow_unknown;

  struct option longopts[] = {
    { "otp-allow-unknown-user",     0, (void*)0L, 'u'},
    { "help",                       0, (void*)0L, 'h'},
    { "help",                       0, (void*)0L, '?'},
    { "otp-db",                     1, (void*)0L, 'o'},
    { "verbose",                    0, (void*)0L, 'v'},
    { "service-name",               0, (void*)0L, 'V'},
    { "otp-challenge-window",       1, (void*)0L, 'w'},
    { "version",                    0, &opt_version, 1},
    { 0, 0, 0, 0},
  };

  /* init xerr */
  xerr_setid(argv[0]);

  otpdb_fname = OTP_DB_FNAME;
  db_flags = 0;
  otp_window = OTP_WINDOW_DEFAULT;
  opt_version = 0;
  otp_allow_unknown = 0;
  service = "otp-openvpn";
  bzero(&ou, sizeof(ou));
  ret = -1; /* fail */

  while ((i = getopt_long(argc, argv, "h?o:uvV:w:", longopts,
    (int*)0L)) != -1) {

    switch (i) {

      case 'h':
      case '?':
        help();
        exit(0); /* be careful here... openvpn treats this as accept auth */

      case 'o' :
        otpdb_fname = optarg;
        break;

      case 'u':
        otp_allow_unknown = 1;
        break;

      case 'v':
        db_flags |= OTP_DB_VERBOSE;
        break;

      case 'V':
        service = optarg;
        break;

      case 'w':
        tmpul = strtoul(optarg, &endptr, 0);
        if (*endptr)
          xerr_errx(1, "stroul(%s): failed at %c.", optarg, *endptr);
        if (tmpul > OTP_WINDOW_MAX)
          xerr_errx(1, "Challenge window %lu > %lu.", tmpul, OTP_WINDOW_MAX);
        otp_window = tmpul;
        break;

      case 0:
        if (opt_version) {
          printf("%s\n", ootp_version);
          exit(0);
        }
        break;

      default:
        xerr_errx(1, "getopt_long(): fatal.");
        break; /* not reached */

     } /* switch */

  } /* while */

  if (!(username = getenv("user")))
    xerr_errx(1, "getenv(user): failed.");

  if (!(pass = getenv("pass")))
    xerr_errx(1, "getenv(pass): failed.");

  if (!(otpctx = otp_db_open(otpdb_fname, db_flags)))
    xerr_errx(1, "otp_db_open(): failed.");

  if ((r = otp_user_exists(otpctx, username)) < 0)
    xerr_errx(1, "otp_user_exists(): failed.");

  if ((r == 1) && (otp_allow_unknown == 0)) {
       
    xerr_info("OTP_AUTH_FAIL via otp_user_exists() allow_unknown=0");
    ret = 1;
    goto done;

  }

  if ((r == 1) && (otp_allow_unknown == 1)) {

    xerr_info("OTP_AUTH_PASS via unknown user and allow_unknown=1");
    ret = 0;
    goto done;

  }

  if (r != 0)
    xerr_errx(1, "User %s does not exist in otp database.", username);

  if (otp_urec_open(otpctx, username, &ou, O_RDONLY,
    FFDB_OP_LOCK_EX) < 0)
    xerr_errx(1, "otp_urec_open(%s): failed.", username);

  if (otp_urec_get(otpctx, &ou) < 0)
    xerr_errx(1, "otp_urec_get(): failed.");

  if (otp_urec_close(otpctx, &ou) < 0)
    xerr_errx(1, "otp_urec_close(): failed.");

  if (ou.flags & OTP_FLAGS_SEND_TOKEN) {
    if (otp_user_send_token(otpctx, username, service) < 0)
      xerr_warnx("otp_user_send_token(): failed.");
  }
  
  if ((r = otp_user_auth(otpctx, username, pass, otp_window)) < 0)
    xerr_errx(1, "otp_user_auth(): failed.");

  if (otp_db_close(otpctx) < 0)
    xerr_errx(1, "otp_user_close(): failed.");

  /* convert otp auth status to openvpn auth return code space */
  if (r == OTP_AUTH_PASS) {
    xerr_info("OTP_AUTH_PASS");
    ret = 0;
  } else {
    xerr_info("OTP_AUTH_FAIL");
    ret = 1;
  }

done:

  return ret;

} /* main.c */

void help()
{
  fprintf(stderr, "otp-ov-plugin [-?hv] [-o otpdb_pathname] [-w otp_window]\n");
  fprintf(stderr, "              -h : help\n");
  fprintf(stderr, "              -v : enable verbose output\n");
} /* help */
